### PR TITLE
[SPARK-31971][WEBUI] Add pagination support for all jobs timeline

### PR DESCRIPTION
**Pull request by sarutak**
### What changes were proposed in this pull request?

This PR proposes to add pagination support for all-jobs timeline.

### Why are the changes needed?

If there are lots of jobs, rendering performance of all jobs timeline can significantly goes down. This issue is reported in SPARK-31967.
For example, after the following operation, UI loading can take >40 sec.
```
(1 to 1000).foreach(_ => sc.parallelize(1 to 10).collect) 
```
Although it's not the fundamental solution, pagination can mitigate the issue.

### Does this PR introduce _any_ user-facing change?

Yes.
Before this change, all displayable jobs are displayed in one page.
all-jobs-shown
After this change, 2 text boxes are added.
One is to set the maximum number of jobs to be displayed in a page and the other is to set  the page number.
`Go` button is also added and if it's pushed, jobs are  re-rendered based on the page size and page number.
all-jobs-pagination

### How was this patch tested?

Additional testcase with the following command.
```
build/mvn -Dguava.version=25.0-jre -Dspark.test.webdriver.chrome.driver=/path/to/chromedriver -Dtest.default.exclude.tags=  -Dtest=none -DwildcardSuites="org.apache.spark.ui.ChromeUISeleniumSuite" test
```

### CODE REVIEW

**Comment by sarutak** [comment]
cc: @gengliangwang 
This improvement is not the fundamental solution for SPARK-31967 but mitigate it so I've filed as an independent issue.

**Comment by gengliangwang** [comment]
@sarutak Thanks for the quick fix. I try on my local setup and it works.

**Comment by gengliangwang** [comment]
image

I think we should make the two pagination sections consistent

**Comment by gengliangwang** [comment]
@sarutak I just confirmed that the issue is caused by #28192 .  The issus can't be reproduced before the change.

**Comment by sarutak** [comment]
@gengliangwang Yes, I've noticed that. The performance issue of `vis-timeline`https://github.com/visjs/vis-timeline/issues/379 happen for newer version of `vis.js`.

**Comment by gengliangwang** [comment]
So, I am actually -1 in this solution, for two reasons:
1. if there are many stages in one job, the perf issue still exists
2. the two paginations in one page seem weird.

I tried a different version 4.21.0 from https://cdnjs.com/libraries/vis
The infinite drawing issue seems better if the zoom is not enabled. 
How about we changing to that version? or, just reverting https://github.com/apache/spark/pull/28192?

**Comment by gengliangwang** [comment]
Let me create a PR for downgrading to 4.21.0 now.

**Comment by sarutak** [comment]
> 1. if there are many stages in one job, the perf issue still exists
> 2. the two paginations in one page seem weird.
For 1), AllJobsPage is not affected even if a job has many stages isn't it? JobPage can be affect but we can apply the same solution. For 2), Yes, it's a little bit weird but the similar solution already exists in StagePage.

**Comment by gengliangwang** [comment]
@sarutak I just created https://github.com/apache/spark/pull/28806.
I believe downgrading the vis version is a safer solution. We need to hotfix this to Spark 3.0 ASAP

**Comment by gengliangwang** [comment]
> Even if no performance issue exists, lots of items rendered in one page can be too noisy.

OK, how about we merge #28806 to 3.0/2.4 first. 
And start the pagination from 3.1 if you insist?

**Comment by sarutak** [comment]
> OK, how about we merge #28806 to 3.0/2.4 first.
> And start the pagination feature from 3.1 if you insist?

I agree. I'll check that PR soon.

**Comment by sarutak** [comment]
> I think we should make the two pagination sections consistent
The upper text fields are for timeline, while the lower ones are for the table so I think they should be independent each other.

**Comment by sarutak** [comment]
retest this please.

**Comment by sarutak** [comment]
Hi @gengliangwang, could you restart discussion?

**Comment by sarutak** [comment]
cc: @dongjoon-hyun

**Comment by gengliangwang** [comment]
retest this please

File: core/src/main/scala/org/apache/spark/ui/jobs/AllJobsPage.scala [hunk 1]
```diff
@@ -184,20 +188,52 @@ private[ui] class AllJobsPage(parent: JobsTab, store: AppStatusStore) extends We
...
+            <div id="application-timeline-zoom-lock">
+              <input type="checkbox"></input>
+              <span>Enable zooming</span>
+            </div>
+            <div>
+              <form id={s"form-event-timeline-page"}
+                    method="get"
+                    action=""
+                    class="form-inline justify-content-end"
+                    style="width: 50%; margin-left: auto; margin-bottom: 0px;">
```
Comment (gengliangwang) [inline core/src/main/scala/org/apache/spark/ui/jobs/AllJobsPage.scala:209]: nit: shall we move the style to webui.css?

File: core/src/main/scala/org/apache/spark/ui/jobs/AllJobsPage.scala [hunk 2]
```diff
@@ -184,20 +188,52 @@ private[ui] class AllJobsPage(parent: JobsTab, store: AppStatusStore) extends We
...
+                <label>Jobs:
+                  {totalJobs}
+                  .
+                  {totalPages}
+                  Pages. Jump to</label>
+                <input type="text"
+                       name="jobs.eventTimelinePageNumber"
+                       id={s"form-event-timeline-page-no"}
```
Comment (gengliangwang) [inline core/src/main/scala/org/apache/spark/ui/jobs/AllJobsPage.scala:217]: Nit: I find similar code in StagePage.scala. But why do we need `{s".."}` here?

File: core/src/main/scala/org/apache/spark/ui/jobs/AllJobsPage.scala (hunk 1 above)
Comment (sarutak) [reply core/src/main/scala/org/apache/spark/ui/jobs/AllJobsPage.scala:209]: Thanks. I'll consider it.

File: core/src/main/scala/org/apache/spark/ui/jobs/AllJobsPage.scala (hunk 2 above)
Comment (sarutak) [reply core/src/main/scala/org/apache/spark/ui/jobs/AllJobsPage.scala:217]: Exactly. I'll remove `s` prefix.

**Comment by gengliangwang** [comment]
@sarutak I am really sorry I missed your ping last week.
The perf problem is resolved in https://github.com/apache/spark/pull/28806. If we want to avoid perf issue in the long term, I think we can avoid rendering the timeline by default. 
I am +0 with the proposal. The pagination makes the visualization simpler, but also makes the UI usage more complex. If you insist, let's merge this one.

**Comment by sarutak** [comment]
> I am +0 with the proposal. The pagination makes the visualization simpler, but also makes the UI usage more complex. If you insist, let's merge this one.

I wondered the pagination feature for task timeline is one compromise so I came up with bringing the idea into jobs timeline but this solution has both upside and downside as you suggested.
So, do you have any better idea to avoid being the timeline noisy?

**Pull request closed without merging**